### PR TITLE
Add gRPC sinks and sources

### PR DESCRIPTION
**Pull request by niruhan**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+                    "but the provided url is " + url + ". " + e.getMessage(), e);
+        }
+        String serviceName = getServiceName(aURL.getPath());
+        this.methodName = getMethodName(aURL.getPath());
+        this.hostPort = aURL.getAuthority();
+        if (optionHolder.isOptionExists(GrpcConstants.CHANNEL_TERMINATION_WAITING_TIME_MILLIS)) {
+            this.channelTerminationWaitingTimeInMillis = Long.parseLong(optionHolder.validateAndGetOption(
+                    GrpcConstants.CHANNEL_TERMINATION_WAITING_TIME_MILLIS).getValue());
+        }
+
```
Comment (ramindu90) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java]: @niruhan lets remove these unnecessary spaces

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static io.siddhi.extension.io.grpc.util.GrpcUtils.getRPCmethodList;
+
+/**
+ * {@code GrpcSink} Handle the gRPC publishing tasks.
+ */
+@Extension(name = "grpc", namespace = "sink",
+        description = "" +
+                "This extension publishes event data encoded into GRPC Classes as defined in the user input " +
```
Comment (ramindu90) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java]: @niruhan shouldn't this be 'encoded event data' ?

File: component/src/test/java/io/siddhi/extension/io/grpc/source/GrpcServiceSourceTestCase.java [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+                        .usePlaintext()
+                        .build();
+                MyServiceGrpc.MyServiceBlockingStub blockingStub = MyServiceGrpc.newBlockingStub(channel);
+                try {
+                        blockingStub.process(request);
+
+                } catch (Exception e) {
+
+                }
+
```
Comment (suhothayan) [inline component/src/test/java/io/siddhi/extension/io/grpc/source/GrpcServiceSourceTestCase.java]: Remove unnessary bank spaces

File: docs/images/siddhi-logo-w.svg [hunk 4]
```diff
@@ -0,0 +1,163 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<!--
+  ~ Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (suhothayan) [inline docs/images/siddhi-logo-w.svg]: remove `docs/api/.md` file 
https://github.com/siddhi-io/siddhi-io-grpc/pull/3/files#diff-9acc5c52455fa613cb9fbb7ee430a3e7

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java (hunk 2 above)
Comment (niruhan) [reply component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java]: No, we encode the data into grpc classes. I.e create objects of grpc classes with the data from events

File: pom.xml [hunk 5]
```diff
@@ -0,0 +1,216 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+~ Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (mohanvive) [inline pom.xml]: Change license year.

Comment (mohanvive) [reply pom.xml]: Check other places as well

File: suppressions.xml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    <!-- do not check swagger generated files -->
+    <suppress checks=".*" files=".*/src/gen/java/.*" />
+
+    <!-- do not check protoc generated files -->
+    <suppress checks=".*" files=".*/src/main/java/org/wso2/grpc/.*" />
+
+    <!-- do not check protoc generated files -->
+    <suppress checks=".*" files=".*/src/test/java/org/wso2/grpc/test/.*" />
+
+</suppressions>
```
[the remark this answers is not in the document]
Comment (niruhan) [reply suppressions.xml]: checkstyle suppression plugin in wso2 parent pom points to https://raw.githubusercontent.com/wso2/code-quality-tools/master/checkstyle/suppressions.xml this file. But we need to add grpc auto-generated files to the suppression. So copied the file locally and added the auto-generated classes.

File: component/pom.xml [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            io.siddhi.extension.io.grpc.*,
+                            org.wso2.grpc.*
+                        </Export-Package>
+                        <Import-Package>
+                            *;resolution:=optional
```
Comment (ramindu90) [inline component/pom.xml]: @niruhan shall we add the exact packages to import?

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+        StreamObserver<Empty> responseObserver = new StreamObserver<Empty>() {
+            //try to send all the siddhi events using one stream observer - impossible without adding client
+            // side streaming in protobuf definition
+            @Override
+            public void onNext(Empty event) {
+            }
+
+            @Override //todo latch based error???
+            public void onError(Throwable t) { //parent method doest have error in its signature. so cant throw
+                // from here
```
Comment (ramindu90) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java]: @niruhan could please remove these unwanted comments

File: component/pom.xml (hunk 7 above)
Comment (niruhan) [reply component/pom.xml]: Fixed in https://github.com/siddhi-io/siddhi-io-grpc/pull/8

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java (hunk 8 above)
Comment (niruhan) [reply component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java]: Fixed in https://github.com/siddhi-io/siddhi-io-grpc/pull/6

**Pull request merged**